### PR TITLE
Fix to query bugs #2033 and #2034:

### PR DESCRIPTION
**Pull request by maumar**
#2033 - invalid results for queries involving null-valued parameters
#2034 - incorrect (missing) parentheses in sql gen for some complex expressions

2033:
We were incorrectly transforming queries that were making comparisons with null-valued parameters.
This was happening because we were performing null semantics expansion first and then getting rid of null-valued parameters comparisons.

Query:
customers.Where(c => c.CustomerName != prm).Select(c => c.Id)

Would get it's comparison expanded to:

SELECT c.Id
FROM Customers AS c
WHERE c.CustomerName <> _prm OR c.CustomerName IS NULL

Which would later be optimized into:
SELECT c.Id
FROM Customers AS c
WHERE c.CustomerName IS NOT NULL OR c.CustomerName IS NULL

which was wrong.

Fix is to perform null-valued parameter comparison pruning first, and only after that perform null semantics expansion.
This means that if we have parameters in the query, null semantics expansion can only be performed once we know the values of parameters (i.e. during sql generation).

If no parameters are used in the query, null semantics is performed early, just like before, so it's results can be cached.

2034:
Problem was that for some cases with multiple AndAlso and OrElse expressions, we would not produce enough parentheses.
Basically we were comparing the node type of the outer binary, with the node type of the inner binary. If they were the same, we would not produce a bracket.
However, if both inner and outer nodes are same (say AndAlso) but the inner node has OrElse inside, it may be required to wrap the inner AndAlso in a bracket (due to the inner OrElse)
This would not happen, and we would produce invalid results.

Fix is to produce parentheses around all binary expressions by default. Some optimizations are still performed - we don't create brackets for binary expressions between two simple terms (column, parameter, constant)
with the exception of logical binary expression for which we always produce parentheses

### CODE REVIEW

**Comment by maumar** [comment]
@anpete

File: src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs [hunk 1]
```diff
@@ -563,16 +575,13 @@ protected override Expression VisitBinaryExpression([NotNull] BinaryExpression b
             }
             else
             {
-                var needClosingParen = false;
+                var needParentheses = !IsSimpleExpression(binaryExpression.Left)
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs]: would it be safe to skip brackets for all non-logical expressions?

**Comment by anpete** [comment]
:shipit:
